route/time_dependent.go

```diff
@@ -0,0 +1,150 @@
+package route
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/nextmv-io/sdk/measure"
+	"github.com/nextmv-io/sdk/model"
+)
+
+// ByIndexAndTime holds a Measure and an EndTime (exclusive) up until this
+// measure is to be used as a Unix timestamp. ByIndexAndTime is to be used with
+// NewTimeDependentMeasure which a slice of ByIndexAndTime.
+type ByIndexAndTime struct {
+	Measure ByIndex
+	EndTime int
+}
+
+// ClientOption can pass options to be used with a TimeDependentMeasure client.
+type ClientOption func(*client)
+
+// TimeDependentMeasuresClient is an interface to handle time dependent
+// measures. It implements a Cost function that takes time into account to
+// calculate costs.
+type TimeDependentMeasuresClient interface {
+	Cost() func(from, to int, data measure.VehicleData) float64
+	DependentByIndex() measure.DependentByIndex
+}
+
+type client struct {
+	measures        []ByIndexAndTime
+	fallbackMeasure ByIndexAndTime
+	cache           sync.Map
+}
+
+// NewTimeDependentMeasuresClient returns a new NewTimeDependentMeasuresClient
+// which implements a cost function.
+// It takes byIndexAndTime measures, where each measure is given with an endTime
+// (exclusive) up until the measure will be used and a fallback measure.
+func NewTimeDependentMeasuresClient(
+	measures []ByIndexAndTime,
+	fallback ByIndex,
+	opts ...ClientOption,
+) (TimeDependentMeasuresClient, error) {
+	sort.SliceStable(measures, func(i, j int) bool {
```

Comment:
I think it would be better to create a local copy of the slice and sort that one instead of the parameter.